### ISSUE TITLE
New custom application stuck in loop

### ISSUE THREAD

**Issue by iaindwyer** [commercetools/merchant-center-application-kit#1439]
We're attempting to create our own custom application, using the starter template provided by CT. We haven't changed any of the code, but on deploying and loading the page in our CT instance we do not see our page as expected. Instead the page starts loading, then a message to contact support briefly appears, and it tries to reload. It keeps doing this until I stop it.

I suspect it's related to the CSP rules. I've copied the errors below (I removed our cloudfront URL, but it does appear in the actual message):

NetworkError: A network error occurred. 39.vendors~app.b4e59cb47ffd1c772784.js:2    Ae https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    httpRequest https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    n https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    sendChunk https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    sendEvents https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    flush https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    flush https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    l https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    reload https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    ForcePageReload https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    Ii https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    Vj https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    unstable_runWithPriority https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    cg https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    Dj https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    Tj https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    X https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    onmessage https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2

—
Second sentry message:
—
NetworkError: A network error occurred. 39.vendors~app.b4e59cb47ffd1c772784.js:2    Ae https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    httpRequest https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    n https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    sendChunk https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    sendEvents https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    flush https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    flush https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    l https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    reload https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    ForcePageReload https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    Ii https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    Vj https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    unstable_runWithPriority https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    cg https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    Dj https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    Tj https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    X https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2    onmessage https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js:2

—
Then a shorter one about the script
—
Content Security Policy: The page’s settings blocked the loading of a resource at inline (“script-src”). onloadwff.js:71:803617

—

Loading failed for the <script> with source “https://<our cloudfront dist>/39.vendors~app.b4e59cb47ffd1c772784.js”. example-starter:1:1

**Comment by shanly-suepaul** [comment]
I've created a repo here that contains everything we've done (without the `dist` and `node_modules` folders): https://github.com/myplanetdigital/commercetools-custom-app-troubleshooting

We:
1. created the app with the `npx` command
2. added the `env.prod.json` and `headers.prod.json` file, as well as the lambda generator
3. ran `yarn build`
4. ran `yarn compile` (alias for the `compile-html` command with the flags specified here: https://docs.commercetools.com/custom-applications/deployment/example-deployment-now-v2)
5. ran `now --prod`
6. Configured the custom app in our CT instance.

And the app is stuck in a CSP loop.

**Comment by emmenko** [comment]
Thanks for the reproduction repo. I found the issue: the `cdnUrl` must be `https`, not `http`.

https://github.com/myplanetdigital/commercetools-custom-app-troubleshooting/blob/1f2f933a59536a4105548020ac77366fa065321d/env.prod.json#L7

This works for me now:

<img width="906" alt="image" src="https://user-images.githubusercontent.com/1110551/79195923-65850300-7e2f-11ea-968c-929bfb78960d.png">

**Comment by iaindwyer** [comment]
I'm a bit confused by something. I've been following the instructions for the AWS deployment where frontEndhost is set to one of the CommerceTools instances, but in this one it looks like it's pointing at the Now deployment.

**Comment by emmenko** [comment]
The `frontendHost` is not an important field, as is only used for Sentry related stuff.

https://github.com/commercetools/merchant-center-application-kit/blob/c73a2c01f50bed30583bae9b052aebfd49d54306/packages/sentry/src/sentry.ts#L32
I think we should

**Comment by emmenko** [comment]
It might be that we will deprecate this field in the future.

**Comment by emmenko** [comment]
@shanly-suepaul @iaindwyer were you able to fix the issue?

**Comment by iaindwyer** [comment]
We haven't had an opportunity to test it yet; I expect we will later today. I should note that the CDN URL is https in our AWS deployment, so I'm not convinced that is the cause of the issue. I did try updating the frontendHost parameter but that didn't change anything.

We've tried deploying to both AWS (my comments) and Now (Shanley's repository) and we're currently getting the same error on both. We'll try updating the Now deployment with the new value you provided and see if that fixes it.

**Comment by emmenko** [comment]
From the reproduction repo you provided, the wrong protocol `http` instead of `https` was the issue.

**Comment by emmenko** [comment]
Let me know when you try it out again if it works or not.

In general, make sure to provide the correct values in the then `env.json` and the `headers.json`, with respect also to the cloud region environment. It's easy to mix things up.

**Comment by shanly-suepaul** [comment]
Thanks for pointing that out @emmenko, and also for your help so far. It didn't fix the redirect problem, unfortunately. We're still seeing CSP errors causing a redirect loop.

**Comment by emmenko** [comment]
Where? The now deployment?

If you use the reproduction repo with the fix, do you still get the errors?

**Comment by shanly-suepaul** [comment]
Yes, I updated the repo at https://github.com/myplanetdigital/commercetools-custom-app-troubleshooting , deployed, and we still get the errors. I tried accessing the MC via the https://mc.us-central1.gcp.commercetools.com and mc.commercetools.co domains and both display the same error. I know I must be missing something relatively simple but I've spent hours poring over the docs and I can't determine what it is.

**Comment by iaindwyer** [comment]
We're actually seeing the same errors in both environments, AWS and Now. The code in the repo Shanly linked is for the Now deployment, but the AWS one is the same.

**Comment by shanly-suepaul** [comment]
Here's a screen recording of what's happening.
[out-3.ogz.zip](https://github.com/commercetools/merchant-center-application-kit/files/4477741/out-3.ogz.zip)

And some screenshots: 
![image](https://user-images.githubusercontent.com/858736/79273756-89693880-7e71-11ea-89db-38d5b944b904.png) 
![image](https://user-images.githubusercontent.com/858736/79273824-9ede6280-7e71-11ea-9fc7-e5de917f84ed.png)

**Comment by emmenko** [comment]
I just tried to use the app hosted at https://mp-ctsc-example-app.now.sh in one of my projects and it works fine.

<img width="752" alt="image" src="https://user-images.githubusercontent.com/1110551/79308629-3aff7c80-7ef9-11ea-860d-8d8b0aac7511.png">

**Comment by tdeekens** [comment]
Sorry, just to get this out of the way: it happens in any browser? Can you clear local storage and cookies for the page? Also try Incognito mode? Sorry, but just to rule this out 🐈.

**Comment by iaindwyer** [comment]
I tried it in Firefox and Chrome, with and without private/incognito mode and observed the same behaviour in all of them.

Are you able to see our CT account and the errors we're getting?

**Comment by shanly-suepaul** [comment]
I have the same problem, after clearing all session/cache/storage and in different browsers. I also tried setting up the custom app in a completely new project and it's exhibiting the same behaviour. Here's a screenshot of the CT config: 
![image](https://user-images.githubusercontent.com/858736/79383446-2dff7f00-7f33-11ea-867b-9be5239f0fb2.png)

**Comment by emmenko** [comment]
@shanly-suepaul I see from the screenshot that you are using `test-custom-app` as the main route path, but as far as I know the custom app is configured to be served at the `examples-starter` route (see my screenshot above https://github.com/commercetools/merchant-center-application-kit/issues/1439#issuecomment-613860031).

**Comment by emmenko** [comment]
This would probably also explain the redirect loop actually.

**Comment by shanly-suepaul** [comment]
@emmenko that was it. I didn't understand that the CT config main route must match the route in the app. Thank you so much for your patience and help. It may be helpful to update the docs to make that connection clear, for people such as myself in the future.

**Comment by emmenko** [comment]
Great, glad that we found the problem.

Regarding the documentation, can you tell me what exactly wasn't clear for you? For instance, we currently have this part documented as:
- https://docs.commercetools.com/custom-applications/getting-started/installing-a-starter-application#renaming-the-application-main-route-path
- https://docs.commercetools.com/custom-applications/register-applications/configuring-a-custom-application
- in the UI **Main route path**

Thanks

**Comment by shanly-suepaul** [comment]
You're right that it discusses the option. Maybe the connection between the two could be made even clearer? For example, in the docs for the MC:

"**Application route path**: The main route path that serves your application. This **must** match a route in your custom application. For example, in the project starter example
the value is [example-starter](https://github.com/commercetools/merchant-center-application-kit/blob/master/application-templates/starter/menu.json)." (adding some language and linking)

The use of "must" would be a flag for me to pay more attention. I misinterpreted this as independent of the routing in the application because I knew that there was some proxying going on behind the scenes. Maybe it's not necessary since I'm clearly the first person to run into this :sweat_smile: 

Again, thanks for your help and feel free to close this issue.

**Comment by emmenko** [comment]
Thanks, I'll definitely take a closer look and try to make it more clear. Thanks a lot for the feedback 🙏
